docs_src/design/adr/device-service/0012-DeviceService-Filters.md

```diff
@@ -0,0 +1,152 @@
+
+- which device sent the event/reading (as determined by the Event device property).
+- the classification or origin (such as temperature or humidity) of data produced by the device as determined by the Reading's name property (which used to be the value descriptor and now refers to the device resource name).
+
+### Two Levels of Device Service Filtering
+
+There are potentially two places where "filtering" in a device service could be useful.  
+
+- One (Sensor Data Filter) - after the device service has communicated with the sensor or device to get sensor values (but before the service creates `Event/Reading` objects and pushes those to core data).  A sensor data filter would allow the device service to essentially ignore some of the raw sensed data.  This would allow for some device service optimization in that the device service would not have perform type transformations and creation of event/reading objects if the data can be eliminated at this early stage.  This first level filtering would, **if put in place**, likely occur in code associated with the read command gets done by the `ProtocolDriver`.
+- Two (Reading Filter) - after the sensor data has been collected and read and put into `Event/Reading` objects, there is a desire to filter some of the `Readings` based on the `Reading` values or `Reading` name (which is the device resource name and formerly the value descriptor name) or some combination of value and name.
+
+At this time, **this design only addresses the need for the second filter (Reading Filter)**.  At the time of this writing, no applicable use case has yet to be defined to warrant the Sensor Data Filter.
+
+### Reading Filters
+Reading filters will allow, not unlike application service filter functions today, to have `Readings` in an `Event` to be removed if:
+
+- the value was outside or inside some range, or the value was greater than, less than or equal to some value
+    - based on the `Reading` value (numeric) of a `Reading` outside a specified range (min/max) described in the device profile for a device resource.  Thus avoiding sending in outlier or jittery data `Readings` that could negatively effect analytics.
+    - based on the `Reading` value (numeric) equal to or near (with in some specified range) the last reading.  This allows a device service to reduce sending in `Event/Readings` that do not represent any significant change.  This differs from the already implemented onChangeOnly in that it is filtering `Readings` within a specified degree of change.
```

Comment:
@tonyespy - the onchange is already there and is not changing.  What is specified here is having a filter that would say something like "if the value is + or - 0.5 from the last reading then filter it" versus being exactly as the last reading filter it.
I will add this to the topics we should clarify in the next device service working group, but since the cache has to already be there to support onChangeOnly, can this feature not be added easily?
In my own use cases, this would be very helpful, especially in values (like temperatures from a decimal pointed reader) that the value will change say 71.8 to 72.2 and I don't really care.  Above or below that and I care.

Reply:
In device-sdk-go the onchange mechanism is only implemented for auto events. The current implementation would need major re-factoring in order to be applied to asynchronous readings and command triggered readings. Also since the onchange flag is actually part of the AutoEvent object, the code would also need to be re-factored to handle your new filter expression(s).